Add support for FHIR Messaging

kjmahalingam opened the pull request:
To test this, run the application as usual, and then POST to `http://127.0.0.1:3000/DSTU2/$process-message` with the contents of `test/fixtures/message.json` in the body. This should drop a row in the database, pulling all the relevant info off of the message and into the appropriate columns.

kli990 commented on index.js:
```diff
@@ -1,9 +1,98 @@
+      .insert([data])
+      .into('trial_data')
+      .then((r) => {
+        return responses.response200;
```
[This](https://www.hl7.org/fhir/DSTU2/messageheader-operations.html) specifies that we should return a bundle

kli990 commented on index.js:
```diff
@@ -1,9 +1,98 @@
+        return responses.response200;
+      })
+      .catch((e) => {
+        return responses.response400;
```
At this point, we might want to return a 500 instead of a 400 status code. Specifically, [this](https://www.hl7.org/fhir/DSTU2/messaging.html#process) says "The client SHALL interpret a 4xx response to indicate that there is no point resubmitting the unaltered message, and a 5xx response to indicate an unexpected error occurred on the part of the server, with the implication that it may be appropriate to resubmit the original message". Right now, we don't really have a way of distinguishing between constraint violations (in which case we do want to return a 400) and generic database errors (in which case we want to return a 500). Until we can do that, I feel like a 500 might be better, but it's up to you

kjmahalingam replied:
Good call. This would be an internal server error, so a 500 makes sense to me!

kli990 commented on test/fixtures/message.json:
```diff
@@ -0,0 +1,414 @@
+            "resource": {
+                "resourceType": "MessageHeader",
+                "id": "6fc53664-8912-4b78-bd43-8d89eea7b651",
+                "timestamp": "2018-08-29T15:31:11-04:00",
```
It looks like we might be dropping the UTC offset in the `capture_time` column. `2018-08-29 15:31:11` is the value I have in my database. We might have to change the column type on that column to `timestamp with time zone` https://www.postgresql.org/docs/12/datatype-datetime.html, and we might also want to have `submission_time` be in the same format just to be consistent.

kli990 commented on utils/responses.js:
```diff
@@ -0,0 +1,42 @@
+      ],
+    },
+  },
+};
```
Really like the idea to move this to a separate file!

kjmahalingam commented on index.js:
```diff
@@ -1,9 +1,98 @@
+  // Finally, try to insert the data into the database
+  return knex
+      .insert([data])
+      .into('trial_data')
```
Good catch, I'm changing it in the POST rather than the table.

kjmahalingam commented on utils/responses.js:
```diff
@@ -0,0 +1,42 @@
+module.exports = {
+  response200: {
+    statusCode: 200,
+    body: {
```
In line with your later comment about how we want to return a Bundle instead of an OperationOutcome, I think we'll want to do the `JSON.stringify` in the lambda itself. We'll have to mutate the response based on the ID of the request message and the time the response message is sent, so I think we should keep it a javascript object until after that point. But I will make sure to return a JSON object by the end!

kjmahalingam commented:
@kli990 I addressed your comments in the latest commit! I tested that it seems to work as expected now. 

Let me know if I missed anything, or messed anything new up!
